src/peer_connection.cpp

```diff
@@ -3064,6 +3064,15 @@ namespace libtorrent
 //			, peer_info_struct(), block_finished.piece_index, block_finished.block_index);
 		picker.mark_as_finished(block_finished, peer_info_struct());
 
+		if (picker.is_piece_ready(block_finished.piece_index))
+		{
+			if (t->alerts().should_post<piece_ready_alert>())
```

Comment:
would it make sense to check ``should_post<>`` first, before ``is_piece_ready()``? The former should be a lot cheaper

Reply:
Yes, that's what I had done initially and changed it alleviate the contention problem over the mutex. I will change it back as we found other ways to deal with that and it probably makes little difference.